### PR TITLE
Publishing many measurements and series at once to c8y

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
Signed-off-by: Pradeep Kumar K J <pradeepkumar.kj@softwareag.com>

## Proposed changes

The code changes include checking the `c8y JSON measurement message size`, If the message size exceeds the threshold size, (16K) then it will not be sent to Cumulocity and an error message will be returned.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_mapper/src/c8y/tests.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@ use c8y_smartrest::{
     error::SMCumulocityMapperError, operations::Operations,
     smartrest_deserializer::SmartRestJwtResponse,
 };
+//use json;
```
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/tests.rs:16]: Remove this.

**Review by rina23q — APPROVED** [review]
LGTM

File: crates/core/tedge_mapper/src/c8y/tests.rs [hunk 2]
```diff
@@ -782,6 +783,183 @@ async fn test_convert_big_event() {
     assert!(converter.convert(&big_event_message).await.is_empty());
 }
 
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+async fn test_convert_big_measurement() {
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/tests.rs:787]: The test cases contain a lot of shared code to create an `http_proxy`, a `converter ` then a `mapper`.

=> I would consider some test helpers.

File: crates/core/tedge_mapper/src/core/error.rs [hunk 3]
```diff
@@ -90,4 +90,7 @@ pub enum ConversionError {
 
     #[error(transparent)]
     FromTimeFormatError(#[from] time::error::Format),
+
+    #[error("Cumulocity mqtt payload threshold(16KB) is exceeded : {0}")]
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/core/error.rs]: Logging 30 chars were not enough to print the payload, it was printing just the type of the payload. I increased it to 50, still does not seem like enough. Shall I increase it more, wont the error message be too long?

File: crates/core/tedge_mapper/src/c8y/tests.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/tests.rs:743]: Moved the creation of the converter to another helper function.

File: crates/core/tedge_mapper/src/c8y/tests.rs (hunk 1 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/tests.rs]: done

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The history of this PR have been messed up by a `git push --force`. Most of the updated files are unrelated. Please, fix that or create another PR.

File: crates/core/tedge_mapper/src/c8y/tests.rs [hunk 4]
```diff
@@ -754,6 +729,145 @@ async fn convert_event_with_extra_fields_to_c8y_json() -> Result<()> {
...
+    buffer.push_str("{");
+    for i in 0..loops {
+        buffer.push_str(&format!("\"temperature\":{}", i));
+        if i != loops - 1 {
+            buffer.push_str(",");
+        }
+    }
+    buffer.push_str("}");
+    buffer
+}
```
Comment (albinsuresh) [inline crates/core/tedge_mapper/src/c8y/tests.rs:868]: Another handy way to create this dummy measurement JSON would be to just create a `HashMap<String, serde_json::Value>` with all these different temperature keys and values and just serialize it at the end with `Value::to_string()`. You could avoid all this special character management to properly insert the `,` , `{` and `}`.

**Review by albinsuresh — CHANGES_REQUESTED** [review]

File: crates/core/tedge_mapper/src/c8y/tests.rs [hunk 5]
```diff
@@ -754,6 +729,145 @@ async fn convert_event_with_extra_fields_to_c8y_json() -> Result<()> {
...
+
+fn create_test_c8y_converter() -> CumulocityConverter<MockC8YHttpProxy> {
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/tests.rs:870]: Yes, some part of the code is duplicated, but the HTTP proxy types are different. So, its not possible to reuse the code.

File: crates/core/tedge_mapper/src/c8y/converter.rs [hunk 6]
```diff
@@ -133,18 +133,31 @@ where
...
+                    vec.push(Message::new(
+                        &self.mapper_config.out_topic,
+                        c8y_json_child_payload,
+                    ));
+                } else {
+                    return Err(ConversionError::C8yMqttPayloadThresholdExceed(
+                        c8y_json_child_payload[0..50].into(),
+                        c8y_json_child_payload.len(),
+                    ));
+                }
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs]: This piece of code is duplicated for the root vs child device cases. But the logic is the same in both case.

Consider to have something along the lines:

```
let c8y_json_payload = match maybe_child_id {
     Some(child_id) => {
        let c8y_json_child_payload =
                    json::from_thin_edge_json_with_child(input.payload_str()?, child_id.as_str())?;
        if !self.children.contains(child_id.as_str()) {
             // ...
        }
        c8y_json_child_payload
     },
     None => json::from_thin_edge_json(input.payload_str()?)?
};

if c8y_json_payload.len() < self.size_threshold.0 {
   // ...
} else {
   // ..
}
```

Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: refactored

File: crates/core/tedge_mapper/src/core/error.rs [hunk 7]
```diff
@@ -90,4 +90,12 @@ pub enum ConversionError {
 
     #[error(transparent)]
     FromTimeFormatError(#[from] time::error::Format),
+
+    #[error("The payload {payload} received on {topic} after translation is {actual_size} greater than the threshold size of {threshold}.")]
```
Comment (albinsuresh) [inline crates/core/tedge_mapper/src/core/error.rs:94]:
```suggestion
    #[error("The size of the mapped payload for the message: {payload} on topic: {topic} is {actual_size}, which is greater than the threshold size of {threshold}.")]
```
```suggestion
    #[error("The payload {payload} received on {topic} after translation is {actual_size} greater than the threshold size of {threshold}.")]
```

**Review by albinsuresh — APPROVED** [review]

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**